agent/sshd/utmp.go

```diff
@@ -0,0 +1,247 @@
+package sshd
+
+/*	The Utmpx struct is derived from the Linux definition (man 5 utmpx).
+
+	At session start, a utmp record is constructed containing the PID,
+	tty name, ID, username, remote host, source IP and time.
+	Type is set to UserProcess and the record is written to UtmpxFile.
+	If a record with the same ID exists, that record is overwritten;
+	otherwise the record is appended to the file.  The same record is
+	appended to WtmpxFile.
+
+	At session end, Type is set to DeadProcess, the User and Host fields
+	are cleared, time is updated and the record is wrtten to UtmpxFile,
+	overwriting the session start record. The same record with ID and
+	source IP address cleared is appended to WtmpxFile.
+*/
+
+import (
+	"os"
+	"unsafe"
+	"bytes"
+	"net"
+	"encoding/binary"
+	"golang.org/x/sys/unix"
+	"github.com/sirupsen/logrus"
+)
+
+type ExitStatus struct {
+	ETermination	int16		// Process temination status - not used
+	EExit		int16		// Process exit status - not used
+}
+
+type Utmpx struct {
+	Type		int16		// UserProcess or DeadProcess
+	Padding		[2]byte		// Padding to align rest of struct
+	Pid		int32		// PID of the ShellHub agent
+	Line		[32]byte	// tty associated with the process
+	ID		[4]byte		// Index, last 4 characters of Line
+	User		[32]byte	// Username
+	Host		[256]byte	// Source IP address
+	Exit		ExitStatus	// Exit status - not used
+	Session		int32		// Session ID - not used
+	Tv		unix.Timeval	// Time entry was made
+	AddrV6		[4]uint32	// Source IP address. IPv4 in AddrV6[0]
+	Reserved	[20]byte
+}
+
+const (
+	UtmpxFile   = "/var/run/utmp"
+	WtmpxFile   = "/var/log/wtmp"
+	UserProcess = 0x7		// Normal process
+	DeadProcess = 0x8		// Terminated process
+)
+
+func utmpStartSession(line, user, remoteAddr string) Utmpx {
+	var u Utmpx
+
+	u.Type = UserProcess
+	a := unix.Timeval{}
+	unix.Gettimeofday(&a)
+	u.Tv.Sec, u.Tv.Usec = a.Sec, a.Usec
+	u.Pid = int32(os.Getpid())
+
+	host, _, err := net.SplitHostPort(remoteAddr)
+
+	if err != nil {
+		logrus.WithFields(logrus.Fields{
+			"ip": remoteAddr,
+		}).Warn("wrong remoteAddr format")
+	} else {
+		ip := net.ParseIP(host)
+		ip4 := ip.To4()
+		if ip4 != nil {
+			u.AddrV6[0] = binary.LittleEndian.Uint32(ip4)
+		} else {
+			ip16 := ip.To16()
+			u.AddrV6[0] = binary.LittleEndian.Uint32(ip16[0:4])
+			u.AddrV6[1] = binary.LittleEndian.Uint32(ip16[4:8])
+			u.AddrV6[2] = binary.LittleEndian.Uint32(ip16[8:12])
+			u.AddrV6[3] = binary.LittleEndian.Uint32(ip16[12:16])
+		}
```

Comment:
Doesn't something like this (https://github.com/openshift/coredns-mdns/blob/d3d4bb1f4a4a0e1619a419a26a3d8dcf2959d990/mdns_test.go#L17-L21) work?

Reply:
I can't see an efficient way of using append to do this.

Reply:
Indeed; I just think we need to add some comment here explaining both code paths. Also, the commit log needs to be fixed as it ended being lost during the revisions.

Reply:
I'm not sure how to fix the commit log. I think I've been doing the commits incorrectly but I'm not clear about how to do it.  I've been doing "git add agent/sshd/utmp.go", "git commit" with a comment, then "git push origin utmp".  I then squash the commits with "git reset --soft HEAD-2", "git commit" and "git push origin +utmp".  I'm still new to git, so would welcome advice on the correct process.

Reply:
Take a look at https://linuxize.com/post/change-git-commit-message/

Otherwise, ping me on Gitter and we do it together.

Reply:
I've done a commit to reset the commit log but without any further code changes.  Further commit to come with minor code changes and additional comments.

Reply:
Commit done with improved comments.